src/Config.php

```diff
@@ -13,134 +14,374 @@
  * isset($config->variableName) or !empty($config->variableName). Attempting to
  * access a nonexistent variable will throw an exception.
  *
+ * // These properties are available at build time and run time.
+ *
  * @property-read string $project
  *   The project ID.
- * @property-read string $environment
- *   The environment ID (usually the Git branch name).
- * @property-read string $application_name
+ * @property-read string $applicationName
  *   The name of the application, as defined in its configuration.
- * @property-read string $tree_id
+ * @property-read string $treeId
  *   An ID identifying the application tree before it was built: a unique hash
  *   is generated based on the contents of the application's files in the
  *   repository.
- * @property-read string $app_dir
+ * @property-read string $appDir
  *   The absolute path to the application.
- * @property-read string $document_root
+ * @property-read string $entropy
+ *   A random string generated for each project, useful for generating has keys.
+ *
+ * // These properties are only available at runtime.
+ *
+ * @property-read string $branch
+ *   The Git branch name.
+ * @property-read string $environment
+ *   The environment ID (usually the Git branch plus a hash).
+ * @property-read string $documentRoot
  *   The absolute path to the web root of the application.
- * @property-read string $mode
- *   The hosting mode (this will only be set on Platform.sh Enterprise, and it
- *   will have the value 'enterprise').
- * @property-read array  $application
- *   The application's configuration, as defined in the .platform.app.yaml file.
- * @property-read array  $relationships
- *   The environment's relationships to other services. The keys are the name of
- *   the relationship (as configured for the application), and the values are
- *   arrays of relationship instances. For example, the hostname of a 'mysql'
- *   relationship may be stored in $config->relationships['mysql'][0]['host'].
- * @property-read array  $routes
- *   The routes configured for the environment.
- * @property-read array  $variables
- *   Custom environment variables.
- * @property-read string $smtp_host
+ * @property-read string $smtpHost
  *   The hostname of the Platform.sh default SMTP server (an empty string if
  *   emails are disabled on the environment).
+ *
  */
 class Config
 {
-    private $config = [];
+
+    /**
+     * Local index of the variables that can be accessed as direct properties (build and runtime).
+     *
+     * The key is the property that will be read.  The value is the environment variable, minus
+     * prefix, that contains the value to look up.
+     *
+     * @var array
+     */
+    protected $directVariables = [
```

Comment:
There's now a mix of `protected` and `private` in the class. We probably don't want to encourage inheritance.

Reply:
I should probably remove the remaining private, then.  In practice I see little value to private in an OSS library and default to protected almost always.